tt_metal/api/buffer.hpp

```diff
@@ -0,0 +1,67 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include "types.hpp"
+#include "tt_metal/impl/buffers/buffer.hpp"
+//==================================================
+//                  BUFFER HANDLING
+//==================================================
+
+namespace tt::tt_metal{
+namespace v1 {
+
+
+/**
+ * @brief Allocates an interleaved DRAM or L1 buffer on the device.
+ *
+ * @param config Configuration for the buffer.
+ * @return Buffer handle to the allocated buffer.
+ */
+Buffer CreateBuffer(const InterleavedBufferConfig &config);
+
+/**
+ * @brief Allocates a buffer on the device.
+ *
+ * @param buffer The buffer to allocate.
+ * @param bottom_up If true, allocates buffer from the bottom up.
+ */
+void AllocateBuffer(Buffer buffer, bool bottom_up);
+
+/**
+ * @brief Deallocates a buffer from the device.
+ *
+ * @param buffer The buffer to deallocate.
+ */
+void DeallocateBuffer(Buffer buffer);
+
+/**
+ * @brief Copies data from a host buffer into the specified device buffer.
+ *
+ * @param buffer Buffer to write data into.
+ * @param host_buffer Host buffer containing data to copy.
+ */
+void WriteToBuffer(Buffer buffer, const std::vector<uint32_t> &host_buffer);
+
+/**
+ * @brief Copies data from a device buffer into a host buffer.
+ *
+ * @param buffer Buffer to read data from.
+ * @param host_buffer Host buffer to copy data into.
+ * @param shard_order If true, reads data in shard order.
+ */
+void ReadFromBuffer(Buffer buffer, std::vector<uint32_t> &host_buffer, bool shard_order = false);
+
+/**
+ * @brief Copies data from a specific shard of a device buffer into a host buffer.
+ *
+ * @param buffer Buffer to read data from.
+ * @param host_buffer Host buffer to copy data into.
+ * @param core_id ID of the core shard to read.
+ */
+void ReadFromShard(Buffer buffer, std::vector<uint32_t> &host_buffer, const uint32_t &core_id);
```

Comment:
`uint32_t` is quite an uncommon data type for a buffer. In most cases it should be `uint8_t` or `std::byte`. I understand this could be referring to the 32B alignment requirement in DRAM access. But still, uncommon and confusing.

Also consider supporting `span` instead of forcing vectors.

Reply:
I agree with you on both points. To my understanding, this alignment is imposed by some other design restriction for now, and this is something that is planned to be relaxed soon so that no alignment is necessary on the host side.

As for `std::span`, I too would love this but for reasons unknown to me, there is a push to support C++17 as a target, and that unfortunately precludes the use of `<span>`. Boost span was also a consideration, but seemed like a very heavy dependency to add for a low-level API. Please understand this is not a consideration that was lightly dismissed.